pkg/clustermesh/internal/remote_cluster.go

```diff
@@ -0,0 +1,377 @@
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/pkg/clustermesh/types"
+	cmtypes "github.com/cilium/cilium/pkg/clustermesh/types"
+	cmutils "github.com/cilium/cilium/pkg/clustermesh/utils"
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+type RemoteCluster interface {
+	Run(ctx context.Context, backend kvstore.BackendOperations, config *types.CiliumClusterConfig) error
+
+	Stop()
+	Remove()
+}
+
+// remoteCluster represents another cluster other than the cluster the agent is
+// running in
+type remoteCluster struct {
+	RemoteCluster
+
+	// name is the name of the cluster
+	name string
+
+	// configPath is the path to the etcd configuration to be used to
+	// connect to the etcd cluster of the remote cluster
+	configPath string
+
+	// clusterSizeDependantInterval allows to calculate intervals based on cluster size.
+	clusterSizeDependantInterval kvstore.ClusterSizeDependantIntervalFunc
+
+	// changed receives an event when the remote cluster configuration has
+	// changed and is closed when the configuration file was removed
+	changed chan bool
+
+	controllers *controller.Manager
+
+	// remoteConnectionControllerName is the name of the backing controller
+	// that maintains the remote connection
+	remoteConnectionControllerName string
+
+	// mutex protects the following variables
+	// - backend
+	// - failures
+	// - lastFailure
+	mutex lock.RWMutex
+
+	// backend is the kvstore backend being used
+	backend kvstore.BackendOperations
+
+	// failures is the number of observed failures
+	failures int
+
+	// lastFailure is the timestamp of the last failure
+	lastFailure time.Time
+
+	metricLastFailureTimestamp prometheus.Gauge
+	metricReadinessStatus      prometheus.Gauge
+	metricTotalFailures        prometheus.Gauge
+}
+
+var (
+	// skipKvstoreConnection skips the etcd connection, used for testing
+	skipKvstoreConnection bool
+)
+
+func (rc *remoteCluster) getLogger() *logrus.Entry {
+	var (
+		status string
+		err    error
+	)
+
+	if rc.backend != nil {
+		status, err = rc.backend.Status()
+	}
+
+	return log.WithFields(logrus.Fields{
+		fieldClusterName:   rc.name,
+		fieldConfig:        rc.configPath,
+		fieldKVStoreStatus: status,
+		fieldKVStoreErr:    err,
+	})
+}
+
+// releaseOldConnection releases the etcd connection to a remote cluster
+func (rc *remoteCluster) releaseOldConnection() {
+	rc.mutex.Lock()
+	backend := rc.backend
+	rc.backend = nil
+	rc.mutex.Unlock()
+
+	// Release resources asynchronously in the background. Many of these
+	// operations may time out if the connection was closed due to an error
+	// condition.
+	go func() {
+		if backend != nil {
+			backend.Close(context.TODO())
```

Comment:
Could we use `context.Background()` here? It seems to be appropriate for async resource release.